test/metabase/test/util.clj

```diff
@@ -806,6 +806,25 @@
            :namespace (name ~collection-namespace))
     (fn [] ~@body)))
 
+(defn do-with-all-users-app-root-permission
+  "Call `f` without arguments in a context where the ''All Users'' group
+  is granted the global app permission `permission`. `permission` can be one of
+  :read or :write."
+  [permission f]
+  (when-not (#{:read :write} permission)
+    (throw (ex-info (str "Unexpected app permission " permission)
+                    {:permission permission})))
+  (tt/with-temp Permissions [_ {:group_id (:id (perms-group/all-users))
+                                :object (cond-> "/collection/namespace/apps/root/"
+                                          (= permission :read) (str "read/"))}]
+    (f)))
+
+(defmacro with-all-users-app-root-permission
+  "Run `body` with the ''All Users'' group being granted the global app
+  permission `permission`. `permission` can be one of :read or :write."
+  [permission & body]
+  `(do-with-all-users-app-root-permission ~permission (fn [] ~@body)))
```

Comment:
This seems a little bespoke... I'm wondering if there's some way we could generalize it a bit so it could be useful in more places in the future. Would something like `with-all-users-permission` that takes a full permissions path work?

```clj
(mt/with-all-users-permission (perms/app-root-collection-permission :read)
  ...)
```

Then you can use it not just here but for all sorts of other things too

Reply:
Implemented.